client/state/utils.js

```diff
@@ -25,6 +25,22 @@ export function isValidStateWithSchema( state, schema, checkForCycles = false, b
 }
 
 /**
+ * Given an action object or thunk, returns an updated object or thunk which
+ * will include additional meta in the action (as provided) when dispatched.
+ *
+ * @param  {(Function|Object)} action Action object or thunk
+ * @param  {Object}            meta   Additional meta to include in action
+ * @return {(Function|Object)}        Augmented action object or thunk
+ */
+export function dispatchWithMeta( action, meta ) {
+	if ( 'function' !== typeof action ) {
+		return { ...action, meta };
+	}
+
+	return ( dispatch ) => action( ( thunkAction ) => dispatch( { ...thunkAction, meta } ) );
```

Comment:
A comment might help considerably here. After an extended pause, I believe that I figured out what this is doing.

```js
// action() takes a dispatch function and does things with it
// this intercepts that passed dispatcher and replaces it with
// a new one which extends the action with the given meta
// before actually dispatching
return dispatch => action( 
	producedAction => dispatch( { 
		...producedAction, 
		meta: { ...get( producedAction, 'meta', {} ), ...meta }
	} )
);
```

Still unclear to me. I think that there is someway to make this more obvious what is happening, though I don't know _how_ to do that yet.